diff --git a/core/lib/Drupal/Core/Entity/Query/Sql/Query.php b/core/lib/Drupal/Core/Entity/Query/Sql/Query.php
index 7adee7f1833..1e45976448d 100644
--- a/core/lib/Drupal/Core/Entity/Query/Sql/Query.php
+++ b/core/lib/Drupal/Core/Entity/Query/Sql/Query.php
@@ -131,16 +131,11 @@ protected function prepare() {
       $this->sqlFields["base_table.$id_field"] = ['base_table', $id_field];
     }

-    // Use max and group by to only return the latest revision in the most
-    // optimal way.
+    // Add a self-join to the base revision table if we're querying only the
+    // latest revisions.
     if ($this->latestRevision && $revision_field) {
-      // Fetch all latest revision ids in a sub-query.
-      $revision_subquery = $this->connection->select($base_table, 'base_table');
-      $revision_subquery->addExpression("MAX(base_table.$revision_field)");
-      $revision_subquery->groupBy("base_table.$id_field");
-
-      // Restrict results only to latest ids.
-      $this->sqlQuery->condition("base_table.$revision_field", $revision_subquery, 'IN');
+       $this->sqlQuery->leftJoin($base_table, 'base_table_2', "[base_table].[$id_field] = [base_table_2].[$id_field] AND [base_table].[$revision_field] < [base_table_2].[$revision_field]");
+      $this->sqlQuery->isNull("base_table_2.$id_field");
     }

     if (is_null($this->accessCheck)) {
